arrows/vxl/average_frames.cxx

```diff
@@ -0,0 +1,788 @@
+    auto pstepA = imA->planestep();
+    auto istepB = imB->istep();
+    auto jstepB = imB->jstep();
+    auto pstepB = imB->planestep();
+    auto istepC = imC->istep();
+    auto jstepC = imC->jstep();
+    auto pstepC = imC->planestep();
+
+    PixType const* planeA = imA->top_left_ptr();
+    PixType const* planeB = imB->top_left_ptr();
+    double*        planeC = imC->top_left_ptr();
+
+    for( unsigned p = 0; p < np;
+         ++p, planeA += pstepA, planeB += pstepB, planeC += pstepC )
+    {
+      PixType const* rowA = planeA;
+      PixType const* rowB = planeB;
+      double*        rowC = planeC;
+
+      for( unsigned j = 0; j < nj;
+           ++j, rowA += jstepA, rowB += jstepB, rowC += jstepC )
+      {
+        PixType const* pixelA = rowA;
+        PixType const* pixelB = rowB;
+        double*        pixelC = rowC;
+
+        for( unsigned i = 0; i < ni;
+             ++i, pixelA += istepA, pixelB += istepB, pixelC += istepC )
+        {
+          *pixelC += scale *
+                     ( static_cast< double >( *pixelB ) -
+                       static_cast< double >( *pixelA ) );
+        }
+      }
+    }
+  }
+
+  // Add to buffer
+  this->window_buffer_.push_back( input );
+  if( this->window_buffer_.size() > this->window_buffer_capacity_ )
+  {
+    this->window_buffer_.pop_front();
+  }
+
+  // Copy into output
+  copy_cast_and_scale( this->last_average_, average, this->should_round_,
+                       1.0 );
+}
+
+// ----------------------------------------------------------------------------
+template < typename PixType >
+void
+windowed_frame_averager< PixType >
+::reset()
+{
+  this->window_buffer_.clear();
+}
+
+// ----------------------------------------------------------------------------
+template < typename PixType >
+size_t
+windowed_frame_averager< PixType >
+::frame_count() const
+{
+  return this->window_buffer_.size();
+}
+
+} // end anonoymous namespace
+
+// ----------------------------------------------------------------------------
+/// Private implementation class
+class average_frames::priv
+{
+public:
+  using frame_averager_byte_sptr =
+    std::unique_ptr< online_frame_averager< vxl_byte > >;
+  using frame_averager_float_sptr =
+    std::unique_ptr< online_frame_averager< double > >;
+
+  priv()
+    : type{ AVERAGER_window }
+      , window_size{ 10 }
+      , exp_weight{ 0.3 }
+      , round{ false }
+      , output_variance{ false }
+      , variance_scale{ 0.0 }
+  {
+  }
+
+  ~priv()
+  {
+  }
+
+  // Internal parameters/settings
+  averager_mode type;
+  unsigned window_size;
+  double exp_weight;
+  bool round;
+  bool output_variance;
+  double variance_scale;
+
+  // The actual frame averager
+  frame_averager_byte_sptr byte_averager;
+  frame_averager_float_sptr float_averager;
+
+  // --------------------------------------------------------------------------
+  // Load model, special optimizations are in place for the byte case
+  void
+  load_model( bool is_byte = true )
```

Comment:
I am struggling to understand the purpose behind only supporting `byte` and `double` as averager types, nor do I see much "optimization" for the byte case. (I guess just because the others all use `double` to average, meaning they have to convert first... but why not just use the actual type?). It seems it would make more sense, instead of having these two specific cases, to have perhaps a `std::[unordered_]map` on the `vil_pixel_format_of` and make `priv::process_frame` a template.

```suggestion
  template < typename T >
  void
  load_model()
```

Note that I believe this would mean that the variance image type also changes from "byte or double" to 'same as the input type'.

Reply:
I agree with you that the specialization for bytes is odd. However, at this point I'd prefer to leave it as-is and address this inefficiency later on a separate PR. Especially because I may have to revisit the variance calculation/usage, I'd like to get this merged in soon.

Reply:
There isn't much optimization in these cases because it wasn't necessary, this process wasn't a speed bottleneck in burnout in the past. The variance method is already used by downstream nodes as-is, if you change them they will break, so if you change the behavior down the liner I would leave in the original and add new as a param.

It is a online variance approximation, given by this paper: R. Ling. Comparison of several algorithms for computing sample
means and variances. Journal of the American Statistical Associa-tion, pages 859–866, 1974.

Reply:
I did this in #1205.